### ISSUE TITLE
multiple instances of a same system

### ISSUE THREAD

**Issue by mikub** [tolitius/yurt#2]
Hi Anatoly,
this is actually more of a question than an issue: is it possible with mount/yurt to spawn n instances of a system (with n presumably changing during run-time)? With Stuart Sierra's component it is possible however I am experiencing some "component fatigue" (I am sure you'd understand) and looking for an alternative. I checked the concept of mount and yurt but so far it seems this is not possible, am I correct?
Basically I would like to have multiple worker systems and start more if I would like to scale or stop some that are already running to do the opposite. Each worker system basically consists of some associated resources (db/mq connections/sessions etc) and a worker thread that can be started/stopped. So I am not looking for testing vs prod system - what I want are multiple instances of the same system with pretty much the same config.

Thanks heaps
Miro

**Comment by tolitius** [comment]
> multiple instances of the same system with pretty much the same config

do you mean within the same JVM?

it is possible with Yurt:

``` clojure
(def bp (yurt/blueprint))
```

``` clojure
(def yurt-one (yurt/build bp))
(def yurt-two (yurt/build bp))
...
(def yurt-n (yurt/build bp))
```

in case you need to plug in different configs to different systems:

``` clojure
(def another-yurt 
  (yurt/build-with bp 
                   {"app.conf/config" another-config}))
```

given that `app.conf/config` is one of the states that hold the configuration.

---

if not within the same JVM, it is possible with both as you can always do another `java -jar -Dconf="conf.edn" app.jar`

about the `pretty much the same config`, in case the app is listening on the same port (or alike), the configs should reflect the change in (in this case) the port.

**Comment by mikub** [comment]
Thanks for quick response, 
I am still not sure I understand how this would work. Briefly looking into the source code of mount/yurt it seems to me that yurt just leverages existing core mount functionality and does not seem to add any one-to-many relation - and if I understand correctly behind the scenes mount defines a var for each state (aka component) and also tracks existing states that were defined via defstate in meta-state atom (which is a map).

If I would start a state (say its start fn would be a function that would start a new thread with some processing loop) three times
`(mount/start #'app/worker)`
`(mount/start #'app/worker)`
`(mount/start #'app/worker)`
what exactly will happen? Will there be three different handles for each of these states? Or will the only var and reference in meta-state atom be overridden every time i start it? I assume the latter, correct?

Using yurt it seems to me that
`(def yurt-one (yurt/build bp))`
`(def yurt-two (yurt/build bp))`
`(def yurt-three (yurt/build bp))`
would basically be the same as calling the code above - it would result into pretty much same situation where I might loose track of overall state as yurt would just swap meta-state with the reference to the last state i started and/or yurt i built, loosing the old state in process. Or am I missing something? 

I understand this is kinda unusual use case, in SS's component world it basically equals to running multiple systems in parallel (some of which may actually have 1..n relationship), which in general is not how majority of ppl even uses component.

**Comment by tolitius** [comment]
> Or am I missing something?

yes :)

## mount

``` clojure
(mount/start #'app/worker)
(mount/start #'app/worker)
(mount/start #'app/worker)
```

would only start `#'app/worker` _once_. all the subsequent calls would check whether it is already started and would not call its `:start` function.

but mount would not be what you want if you are looking to start multiple systems from the same components _within the same JVM_.

## yurt

``` clojure
(def yurt-one (yurt/build bp))
(def yurt-two (yurt/build bp))
(def yurt-three (yurt/build bp))
```

what yurt does is to:
- read the blueprint provided by mount
- start the system using that blueprint
- _detach_ these states from vars

so "vars", in case of yurt, serve as a guidance rather than a runtime use. 

In other words: once one system is started with `yurt/build`, another system is ready to start from a clean slate. The first system would still be there decoupled (i.e. detached) from any future systems.

## try it

yurt comes with a [sample app](https://github.com/tolitius/yurt/tree/master/dev/clj/neo).

just clone yurt, cd to it:

``` bash
$ git clone https://github.com/tolitius/yurt
$ cd yurt
```

and try it.

Start two separate system from the same exact blueprint (sources). [here is how](https://github.com/tolitius/yurt#show-me)

**Comment by mikub** [comment]
ok, got it, thanks! will definitely give it a try.
btw looking at the detaching functionality should i be wary of any race conditions if e.g. two concurrent threads are trying to build a yurt at the same time? I did not see any locks there so am presuming it will only serialize on the individual vars, but not on the yurt as a whole, correct? I should therefore probably serialize any access to yurt/build fn in my app, right?

**Comment by tolitius** [comment]
> two concurrent threads are trying to build a yurt at the same time

right, good catch. This is currently not supported. i.e. I would expect systems to start "sequentially".

sure, try it. let me know whether you have any more questions
